keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,570 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, a Kubernetes Service accepts a list of ports to be exposed by it.
+It is possible to specify any number of ports (as long as the service is within the max object size limit), by listing them in the service spec.
+This can be tedious if the service needs a large number of ports.
+This KEP proposes to add a new field to the Service spec to allow exposing the entire port range(1 to 65535).
+
+## Motivation
+
+There are several applications like SIP/RTP/Gaming servers that need a lot(1000+) of ports to run multiple calls or media streams.
+Currently the only option is to specify every port in the Service Spec. A request for port ranges in Services has been open in https://github.com/kubernetes/kubernetes/issues/23864. Implementing port ranges are challenging since iptables/ipvs do not support remapping port ranges. Also, in order to specify several non-contiguous port ranges, the user will have to expose the entire valid port range. Hence, this proposal to set a single field in order to expose the entire port range and implement the service clients and endpoints accordingly.
+[A survey](https://docs.google.com/forms/d/1FOOG2ZoQsnJLYAjnhEtSPYmUULWFNe88iXR7gtFcP7g/edit) was sent out to collect the use-cases for AllPorts support - [results.](http://tiny.cc/allportsslides)
+
+### Goals
+
+* Allow users to optionally expose the entire port range via a Service (of Type LoadBalancer or ClusterIP).
+
+### Non-Goals
+
+* Supporting Port Ranges in a Service.
+* Changing the default behavior of Service ports.
+
+## Proposal
+
+The proposal here is to introduce an `allPorts` boolean(*bool) field to the [service API.](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#servicespec-v1-core)
+A service that sets this field to true will be reachable on any valid port [1 to 65535]. The backend pods have to be configured accordingly. The backend pods will receive requests with the same port number - in other words, port remapping will not be possible.
+
+The [APIServer validation](https://github.com/kubernetes/kubernetes/blob/b9ce4ac212d150212485fa29d62a2fbd783a57b0/pkg/apis/core/validation/validation.go#L4162) that disallows empty `ports` will be relaxed, when `allPorts` is set.
+
+The value of `allPorts` field can be toggled on supported service types.
+
+### Supported service types
+
+Setting this field will be supported for:
+
+* ServiceType=ClusterIP
+* ServiceType=LoadBalancer
```

Comment:
How will this work with cloud load balancers that only forward to an assigned port and / or only listen on an assigned port?